lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,376 @@
+      doc: """
+      Determines random number generation for centroid initialization.
+      If the seed is not provided, it is set to `System.system_time()`.
+      """
+    ]
+  ]
+
+  @doc """
+  Cluster the dataset using affinity propagation.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:affinity_matrix` - Affinity matrix. It is a negated squared euclidean distance of each pair of points.
+
+    * `:clusters_centers` - Cluster centers from the initial data.
+
+    * `:cluster_centers_indices` - Indices of cluster centers.
+
+    * `:num_clusters` - Number of clusters.
+
+    * `:similarity_matrix` - Similarity matrix. It is similarity of each pair of points.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,-5,7,32], [-1,3,6,1], [1,-2,5,2]])
+      iex> Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      %Scholar.Cluster.AffinityPropagation{
+        labels: Nx.tensor([0, 3, 3, 3]),
+        cluster_centers_indices: Nx.tensor([0, -1, -1, 3]),
+        affinity_matrix: Nx.tensor(
+          [
+            [-0.0, -6162.0, -5358.0, -5499.0],
+            [-6162.0, -0.0, -1030.0, -913.0],
+            [-5358.0, -1030.0, -0.0, -31.0],
+            [-5499.0, -913.0, -31.0, -0.0]
+          ]),
+        cluster_centers: Nx.tensor(
+          [
+            [12.0, 5.0, 78.0, 2.0],
+            [:nan, :nan, :nan, :nan],
+            [:nan, :nan, :nan, :nan],
+            [1.0, -2.0, 5.0, 2.0]
+          ]
+        ),
+        num_clusters: Nx.tensor(2, type: :u64),
+        similarity_matrix: Nx.tensor(
+          [
+            [-971.5, -6162.0, -5358.0, -5499.0],
+            [-6162.0, -971.5, -1030.0, -913.0],
+            [-5358.0, -1030.0, -971.5, -31.0],
+            [-5499.0, -913.0, -31.0, -971.5]
+          ]
+        )
+      }
+  """
+  deftransform fit(data, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    opts = Keyword.update(opts, :self_preference, false, fn x -> x end)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(data, seed, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp fit_n(data, seed, opts) do
+    data = to_float(data)
+    iterations = opts[:iterations]
+    damping_factor = opts[:damping_factor]
+    self_preference = opts[:self_preference]
+
+    {initial_a, initial_r, s, affinity_matrix} =
+      initialize_matrices(data, self_preference: self_preference)
+
+    {n, _} = Nx.shape(initial_a)
+    key = Nx.Random.key(seed)
+    {normal, _new_key} = Nx.Random.normal(key, 0, 1, shape: {n, n}, type: Nx.type(s))
+
+    s =
+      s +
+        normal *
+          (Nx.Constants.smallest_positive_normal(Nx.type(s)) * 100 +
+             Nx.Constants.epsilon(Nx.type(data)) / 10 * s)
+
+    range = Nx.iota({n})
+
+    {a, r, _, _, _} =
+      while {a = initial_a, r = initial_r, s = s, range = range, i = 0},
```

Comment:
I think this also works:

```suggestion
      while {a = initial_a, r = initial_r, s, range, i = 0},
```